static/app/components/forms/teamSelector.tsx

```diff
@@ -0,0 +1,264 @@
+import * as React from 'react';
+import styled from '@emotion/styled';
+
+import {addTeamToProject} from 'app/actionCreators/projects';
+import {Client} from 'app/api';
+import Button from 'app/components/button';
+import SelectControl, {ControlProps} from 'app/components/forms/selectControl';
+import {IconAdd, IconUser} from 'app/icons';
+import {t} from 'app/locale';
+import space from 'app/styles/space';
+import {Organization, Project, Team} from 'app/types';
+import withApi from 'app/utils/withApi';
+import withOrganization from 'app/utils/withOrganization';
+import withTeams from 'app/utils/withTeams';
+
+import IdBadge from '../idBadge';
+import Tooltip from '../tooltip';
+
+const UnassignedWrapper = styled('div')`
+  display: flex;
+  align-items: center;
+`;
+
+const StyledIconUser = styled(IconUser)`
+  margin-left: ${space(0.25)};
+  margin-right: ${space(1)};
+  color: ${p => p.theme.gray400};
+`;
+
+// An option to be unassigned on the team dropdown
+const unassignedOption = {
+  value: null,
+  label: (
+    <UnassignedWrapper>
+      <StyledIconUser size="20px" />
+      {t('Unassigned')}
+    </UnassignedWrapper>
+  ),
+  searchKey: 'unassigned',
+  actor: null,
+  disabled: false,
+};
+
+// Ensures that the svg icon is white when selected
+const unassignedSelectStyles = {
+  option: (provided, state: any) => ({
+    ...provided,
+    svg: {
+      color: state.isSelected && state.theme.white,
+    },
+  }),
+};
+
+type Props = {
+  api: Client;
+  organization: Organization;
+  teams: Team[];
+  onChange: (value: any) => any;
+  /**
+   * Function to control whether a team should be shown in the dropdown
+   */
+  teamFilter?: (team: Team) => boolean;
+  /**
+   * Can be used to restrict teams to a certain project and allow for new teams to be add to that project
+   */
+  project?: Project;
+  /**
+   * Controls whether the value in the dropdown is a team id or team slug
+   */
+  useId?: boolean;
+  includeUnassigned?: boolean;
+} & ControlProps;
+
+type TeamActor = {
+  type: 'team';
+  id: string;
+  name: string;
+};
+
+type SelectableTeam = {
+  value: string;
+  label: React.ReactElement;
+  searchKey: string;
+  actor: TeamActor;
+  disabled?: boolean;
+};
+
+type TeamOption = SelectableTeam | typeof unassignedOption;
+
+type State = {
+  options: TeamOption[];
+};
+
+class TeamSelector extends React.Component<Props, State> {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      options: this.teamOptions,
+    };
+  }
+
+  // TODO(ts) This type could be improved when react-select types are better.
+  selectRef = React.createRef<any>();
+
+  get teamOptions() {
+    const {teams, teamFilter, includeUnassigned, project} = this.props;
+    const filteredTeams = teamFilter ? teams.filter(teamFilter) : teams;
+    if (project) {
+      const teamsInProjectIdSet = new Set(project.teams.map(({id}) => id));
```

Comment:
tiny nit, but this can be code-golfed to

```tsx
      const teamsInProjectIdSet = new Set(project.teams.map(t => t.id));
```

Arguably marginally easier to read as well